sympy/solvers/ode/single.py

```diff
@@ -2695,7 +2700,381 @@ def _get_general_solution(self, *, simplify_flag: bool = True):
+      can do ``assert simplify(wronskian(sollist)) != 0`` to check for linear
+      independence.  Also, ``assert len(sollist) == order`` will need to pass.
+    - ``returns = 'both'``, return a dictionary ``{'sol': <solution to ODE>,
+      'list': <list of linearly independent solutions>}``.
+
+    Examples
+    ========
+
+    >>> from sympy import Function, dsolve, pprint
+    >>> from sympy.abc import x
+    >>> f = Function('f')
+    >>> eq = f(x).diff(x, 2)*x**2 - 4*f(x).diff(x)*x + 6*f(x)
+    >>> pprint(dsolve(eq, f(x),
+    ... hint='nth_linear_euler_eq_homogeneous'))
+            2
+    f(x) = x *(C1 + C2*x)
+
+    References
+    ==========
+
+    - https://en.wikipedia.org/wiki/Cauchy%E2%80%93Euler_equation
+    - C. Bender & S. Orszag, "Advanced Mathematical Methods for Scientists and
+      Engineers", Springer 1999, pp. 12
+
+    # indirect doctest
+
+    """
+    hint = "nth_linear_euler_eq_homogeneous"
+    has_integral = False
+
+    def _matches(self):
+        eq = self.ode_problem.eq_preprocessed
+        f = self.ode_problem.func.func
+        order = self.ode_problem.order
+        x = self.ode_problem.sym
+        match = self.ode_problem.get_linear_coefficients(eq, f(x), order)
+        self.r = None
+        does_match = False
+
+        if order and match:
+            coeff = match[order]
+            factor = x**order / coeff
+            self.r = {i: factor*match[i] for i in match}
+        if self.r and not any(not self._test_term(self.r[i], i) for i in
+                self.r if i >= 0):
+            if not self.r[-1]:
+                does_match = True
+
+        return does_match
+
+    def _test_term(self,coeff, order):
+        r"""
+        Linear Euler ODEs have the form  K*x**order*diff(y(x),x,order) = F(x),
+        where K is independent of x and y(x), order>= 0.
+        So we need to check that for each term, coeff == K*x**order from
+        some K.  We have a few cases, since coeff may have several
+        different types.
+        """
+        x = self.ode_problem.sym
+        f = self.ode_problem.func.func
+        if order < 0:
+            raise ValueError("order should be greater than 0")
+        if coeff == 0:
+            return True
+        if order == 0:
+            if x in coeff.free_symbols:
+                return False
+            return True
+        if coeff.is_Mul:
+            if coeff.has(f(x)):
+                return False
+            return x**order in coeff.args
+        elif coeff.is_Pow:
+            return coeff.as_base_exp() == (x, order)
+        elif order == 1:
+            return x == coeff
+        return False
+
+    def _get_sols(self,r):
+        x = self.ode_problem.sym
+        f = self.ode_problem.func.func
+
+        # First, set up characteristic equation.
+        chareq, symbol = S.Zero, Dummy('x')
+
+        for i in r.keys():
+            if not isinstance(i, str) and i >= 0:
```

Comment:
If `r` is a dict then it isn't necessary to call `keys` here (the dict is already iterable).

Why would `i` be a `str`?

Reply:
`r` can have key named 'trialset' which will be used for undetermined coefficient.

Reply:
It would be better to store trialset as a separate attribute rather than mix it up with the roots:
```
self.roots = ...
self.trialset = ...
```